Allow loopback interfaces on all node types when specified by the user

ipspace opened the pull request:
This is the main logic for fixing #2261 and #2266. @jbemmel: would appreciate if you could double-check it.

Next, I have to create test cases to validate the edge conditions (**loopback.ipv4** on host, **loopback: True** and **loopback: False** work correctly), and finally run all known devices through these test cases to verify that the configuration templates don't break.

jbemmel commented on netsim/roles/router.py:
```diff
@@ -12,44 +12,5 @@
 
 from . import select_nodes_by_role
```
No longer used :)

jbemmel commented on netsim/augment/nodes.py:
```diff
@@ -194,6 +194,61 @@ def check_duplicate_mgmt_addr(topology: Box) -> None:
+    n.loopback = {}                               # ... we can do that, will start with no extra data
+  elif lb_value is None:                          # Oh, the user doesn't care
+    if n.get('role','router') != 'router':        # ... so they will not get a loopback for a non-router device
+      return
```
What about ```loopback: bla```?

jbemmel commented on netsim/augment/links.py:
```diff
@@ -1077,15 +1077,11 @@ def copy_link_gateway(link: Box, nodes: Box) -> None:
 """
 def set_node_af(nodes: Box) -> None:
   for n in nodes.values():
     for af in ['ipv4','ipv6']:
```
```for af in ['ipv4','ipv6'] if af not in n.af```

ipspace replied:
Yeah, we could do that optimization. Won't buy us much, but it's nice to have ;)

jbemmel reviewed (changes requested):
Couple of corner cases

ipspace commented on netsim/augment/nodes.py:
```diff
@@ -194,6 +194,61 @@ def check_duplicate_mgmt_addr(topology: Box) -> None:
+        f"Address pool {pool} cannot contain unnumbered/LLA addresses",
+        category=log.IncorrectType,
+        module='nodes')
+    elif not n.loopback[af] and not (prefix_list[af] is False):
```
Thank you, will check and add to a test case

ipspace commented:
@jbemmel: Addressed all your corner cases (thank you), made loopback parameter processing as flexible as I possibly could, documented the changes, fixed the templates that broke without the loopback interface...

You might want to look at this again before I merge it.

jbemmel commented on tests/topology/expected/dual-stack.yml:
```diff
@@ -238,7 +238,6 @@ nodes:
     loopback:
       ifindex: 0
       ifname: Loopback0
-      ipv4: 10.0.0.1/32
```
Is this expected? Topology has dual-stack loopback pool

jbemmel commented on tests/topology/expected/unnumbered.yml:
```diff
@@ -36,12 +36,10 @@ links:
   - ifindex: 2
     ifname: Ethernet1/2
     ipv4: true
-    ipv6: true
```
Expected?

jbemmel reviewed (approved):
looks ok to me
